crypto-keys/src/main/java/com/palantir/crypto2/keys/serialization/SymmetricKeySerializerV3.java

```diff
@@ -0,0 +1,143 @@
+/*
+ * (c) Copyright 2018 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.crypto2.keys.serialization;
+
+import com.google.common.base.Preconditions;
+import com.google.common.base.Throwables;
+import com.palantir.crypto2.keys.KeyMaterial;
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.security.InvalidAlgorithmParameterException;
+import java.security.InvalidKeyException;
+import java.security.NoSuchAlgorithmException;
+import javax.crypto.Cipher;
+import javax.crypto.IllegalBlockSizeException;
+import javax.crypto.NoSuchPaddingException;
+import javax.crypto.SecretKey;
+import javax.crypto.spec.IvParameterSpec;
+
+/**
+ * Serializer for wrapping and unwrapping {@link KeyMaterial}. The {@link #wrap} method returns the KeyMaterial
+ * serialized as follows, which is the same format the {@link #unwrap} method expects:
+ *
+ * <pre>
+ *  +--------------------------------------------------------------+
+ *  | version | wrapping iv | key algorithm length | key algorithm |
+ *  |   byte  |    byte[]   |          int         |     byte[]    |
+ *  +--------------------------------------------------------------+
+ *  +-------------------------------------------------------+
+ *  | wrapped key length | wrapped key | iv length |   iv   |
+ *  |        int         |    byte[]   |    int    | byte[] |
+ *  +-------------------------------------------------------+
+ * </pre>
+ */
+enum SymmetricKeySerializerV3 implements SymmetricKeySerializer {
+    INSTANCE;
+
+    private static final String AES_CTR_NO_PADDING = "AES/CTR/NoPadding";
+    private static final int IV_SIZE = 16;
+
+    static final int VERSION = 3;
+
+    @Override
+    public byte[] wrap(KeyMaterial keyMaterial, SecretKey key) {
+        ByteArrayOutputStream byteStream = new ByteArrayOutputStream();
+        DataOutputStream stream = new DataOutputStream(byteStream);
+
+        byte[] wrappingIv = KeyMaterials.generateIv(IV_SIZE);
+        Cipher keyWrappingCipher = getCipher(Cipher.WRAP_MODE, key, wrappingIv);
+        SecretKey secretKey = keyMaterial.getSecretKey();
+
+        try {
+            stream.write(VERSION);
+
+            stream.write(wrappingIv);
```

Comment:
realize its constant, but should we write the IV length before its bytes?

Reply:
didn't include that in this case given this IV is controlled by this wrapping function and not provided by the caller (which is the case with the other IV). think if we want to change the IV length or the algorithm, it's nicer to just bump the version we're writing rather than making this format more generic.

Reply:
+1

Reply:
fine by me